Improve `global` detection

penguinencounter opened the pull request:
resolves #112

- [x] Integrate `---@plugin`
- [ ]

penguinencounter committed:
Add minimal Lua string and comment parser

penguinencounter committed:
Fix using vars from wrong scope

penguinencounter committed:
Go fast

penguinencounter committed:
only show perf if slow

penguinencounter committed:
go Faster

penguinencounter committed:
Fix the Lexer Broken Detector

penguinencounter committed:
handle edge case of string at EOF

penguinencounter committed:
Fix the comment clobber bug

JanSharp commented:
I was aware of this issue however I've not added parsing for strings because it'd be even more performance cost, and the plugin is already performance heavy. I just added single line comment detection to each module that inserts newlines.

With #110 you could fix `global` being used "as an expression" inside of a s tring literal by adding `---@plugin disable-next-line: global` or similar.

Looking at your implementation there are some issues, such as table allocations that can be removed as well as an O(n) operation that can be changed to O(log(n)) using binary search, which can then be improved further by remembering the lower bound making it even closer to O(1) in most cases. This is noticable with files with many comments, string literals and `global` usage, a 1.5k line file with ~175 `global`, ~246 comments and ~237 string literals goes from 59 ms to 667 ms.

_However_ by fixing the issues mentioned above this is actually promising because it could replace the single line comment check from every file and they could all use the same data. This ignore logic would actually be very similar to what I've done in #110.

So what I'd like to do, if it's alright with you, is extend upon your lexer, imporving performance and integrate it with all the modules, potentially combining it with the data structure in #110. After that I'd compare performance with and without the lexer and we'll see if it's good enough to be kept. If it isn't, then you'll have to use `---@plugin disable...` once it's released.

(I'll also go ahead and check if there's any other locations where there's a missing space before inserted `--`)

penguinencounter commented:
@JanSharp Thanks for looking! For the performance changes, would you rather

* I close this PR so you can open a new one
* I add you as a contributor to my fork
* something else?

Either way, you can go ahead with the changes :)

penguinencounter commented on luals-addon/factorio/factorio-plugin/global.lua:
```diff
@@ -89,12 +89,20 @@ local function replace(uri, text, diffs)
     for preceding_text, start, finish, ignore_pos, ignore_char, final_pos in
       util.gmatch_at_start_of_line(text, "([^\n]-)%f[a-zA-Z0-9_]()global()[^%S\n]*()([=.%[]?)()")--[[@as fun(): string, integer, integer, integer, string, integer]]
     do
+      for _, non_code in ipairs(non_code_positions) do
+        if non_code.from <= start and non_code.to >= finish then
+          goto continue
+        end
+      end
```
This appears to be `O(NM)` worst case. By https://github.com/justarandomgeek/vscode-factoriomod-debug/pull/113#issuecomment-1890889081, this should probably be replaced with a binary search of some sort

penguinencounter committed:
Remove unused function

penguinencounter commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
+  end
+
+  local function append_range()
+    ranges[#ranges + 1] = {
+      from = start,
+      to = _end,
+      length = _end - start + 1,
+      content = source:sub(start, _end),
+    }
```
This could be turned into an array: `{start, _end}`
(or maybe just have `ranges` be `{start1, end1, start2, end2, ..., startN, endN}`)

`length` and `content` are not actually used anywhere outside of testing and might cause slowdown with interning.

JanSharp replied:
Yep, it would be the latter. Reason being that the latter does not do a heap allocation (table construction) for every range, it's just 1 big table. And that table can remain large, so by reusing the same table throughout the lifetime of the entire plugin it cuts down on allocations significantly. And allocations are as per usual the expensive part in programming.

penguinencounter commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = 0
+    while take("=") do
+      count = count + 1
+    end
```
Is `string.match` faster here?

penguinencounter commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
+      short_string = function()
+        -- we still need to handle escapes correctly
+        if not char_escaped then
+          cursor = string.match(source, patterned_delimit, cursor) or #source + 1
```
For this and the next 3 instances of `string.match`, `string.find` wouldn't require the position capture `()`, but is it faster?

JanSharp replied:
I've tested it a bit, there's no measurable difference. So :shrug: other parts are more expensive, basically.

penguinencounter commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
+    local origin = cursor
+
+    ---@type {[LexerState]: fun()}
+    local modes = {
```
Maybe construct this table and `default` outside of loop, or outside the function scope entirely? Not sure if this gets optimized by `luac`.
Alternatively, use a `if`/`elseif`/`else` set instead?

JanSharp replied:
I actually don't know what exact version of Lua the Lua LS is using, but it seems to be some kind of Lua 5.4. Lua does not do the kind of optimization you mentioned here, so yes the table constructor would have to be moved out of the loop/function. Lua in general has a one pass compiler, so the only optimizations it does is some bare minimum constant folding for things like `1 + 1`. LuaJIT is a different story, but I'm pretty sure Lua LS does not use LuaJIT

penguinencounter reviewed (commented):
Performance comments

JanSharp commented:
Sicne I can't bulk reply to the other performance comments I'll do it here: I've had the same thoughts, I think the `while take("=") do` is probably faster with a `string.match`, because it's fewer function calls and I also thought about string interning. Speaking of, I also thought about not using `string.sub` for any of the parsing for that same reason.

For the other ones, I'm not sure if `string.match` or `string.find` is faster, something I wanted to test just to make sure. The difference is most likely pretty minor though.

> @JanSharp Thanks for looking! For the performance changes, would you rather
> 
> * I close this PR so you can open a new one
> * I add you as a contributor to my fork
> * something else?
> 
> Either way, you can go ahead with the changes :)

Option 2 is probably the cleanest/straight forward, if that's alright with you.

And also thank you in advance for the PR. Like I mentioned previously I can't guarantee if it'll be fast enough, but you've set the ground works for actually figuring it out, where I previously simply assumed it would never be close.

JanSharp commented:
I just realized that I had the debugger attached when testing performance earlier today. The same file mentioned previously actually went from 27.2 ms to 33.5 ms. Oops, my bad.
And fyi I'm testing performance by running it inside of Factorio, so with Factorio Lua. Simply because I'm already using Factorio for plugin dev anyway (to actually view generated diffs). I'll have to do a least some tests in the real plugin environment :see_no_evil:

JanSharp committed:
Merge branch 'next' into global-is-not-a-string-or-comment

JanSharp committed:
Remove timer

Specifically because there are no runtime timers for any part of the
  plugin at this time. Having a timer for a single part of the plugin
	would not provide useful information to neither the user nor the
	programmer when receiving reports. I may add timers to all parts of
	the plugin eventually, but at the moment I'm happy with measuring
	performance locally (and using the Factorio Profiler) and simply
	fixing issues as they come up

JanSharp committed:
Change code ranges to use binary search

Change to reuse the same ranges table throughout the lifetime of the
  entire plugin
Note that this is using a copy pasted binary search function that is
  used for plugin disable. The idea is that those 2 data structures will
	get merged into one and the copy pasted function will get removed

JanSharp committed:
Improve block string/comment `=` counting

penguinencounter committed:
Move modes table constructor outside of loop

JanSharp commented:
Oh I already did that, I just didn't push the commits, sorry. I've got a habbit of not pushing commits unfortunately.

penguinencounter commented:
oh oops
shall I move the branch back, or have you already merged it on your side

JanSharp commented:
I can merge it no worries. It's better this way anyway so you get credit for it in the git blame.
(I felt bad for moving that code making it look like I touched all of it in git history even though you did the work)

JanSharp committed:
Merge remote-tracking branch 'penguinencounter/global-is-not-a-string-or-comment' into global-is-not-a-string-or-comment

JanSharp commented:
Huh github doesn't show the other commits in the list in the PR, but you can look at them through other means. Either way, feel free to push more changes if you want to.

My next step is going to be moving the `---@plugin` parsing into this parser actually, which is a pretty big change, but I realized that it's the simplest and presumably an efficient solution to having both `---@plugin` disabling and string/comment disabling co-exist.

penguinencounter commented:
@JanSharp the commits are there, they're just earlier in the list because you committed them locally yesterday

JanSharp commented:
Oh dear, I scrolled up because I thought that was the case... yet I missed them. :sweat_smile:

penguinencounter commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -368,6 +374,221 @@ local function add_chain_diff(chain_diff, diffs)
+  ---@return boolean, integer | nil
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = source:match("^=*()", cursor) - cursor
+    cursor = cursor + count
+    if take("[") then
+      return true, count
+    else
+      return false, nil
+    end
```
would this work better?
```suggestion
    local match = source:match("^=*%[", cursor)
    if not match then return false, nil end
    cursor = cursor + #match
    return true, #match - 1
```

JanSharp replied:
Yea it would, would be better tbh. Though something I have tested in the past is that having a match return a string vs retunring a number is slower, so I'd keep the `()` at the end to calculate count and to update cursor in any way that works.
Although saying that I did test performance before and after moving the modes switch table outside of the loop and the difference was within margin of error (so there was no difference), so realistically the suggested change is good as is.

penguinencounter committed:
Use pattern to match brackets

penguinencounter committed:
reset first code range start when loading multiple files

penguinencounter committed:
whoops lua

JanSharp commented on luals-addon/factorio/factorio-plugin/util.lua:
```diff
@@ -390,6 +390,7 @@ function clean_up_code_ranges()
   for i = code_ranges_count, 2, -1 do
     code_ranges[i] = nil
   end
+  code_ranges[0] = 0
```
There's no need to reset `code_ranges[1]` to `0` because the loop above loops from `code_ranges_count` down to index `2`, so index `1` remains untouched. At the end of the day it's a code readability situation where I prefer having the loop not touch `1` to make it clear that index `1` is truly static.
I should add a comment why that first element must always exist though I just noticed.

JanSharp committed:
Add comment for how `code_ranges` int array works

JanSharp committed:
Handle malformed short/quoted strings

The plugin must handle invalid code states. One might question why
  someone would have an unterminated quoted string and continue typing
  or inspecting the lines of code past it, but they totally might.
  Additionally by not detecting unterminated strings at the end of line
  it could flip the ranges which are considered code or strings, which
  could cause the plugin to generate more syntax errors than there
  actually are in the code

JanSharp committed:
Add "early" return for clarity

JanSharp committed:
Remove unrelated code added in 30d10b2

Dang it, this is why you should never push immediately after committing
  You just end up with 2 fix commits instead of amending the original
  commit

JanSharp committed:
Move pre and post process functions to the bottom

Similar to how it makes more sense to put script.on_event handlers at
  the bottom of the file in Factorio mods, these functions make more
  sense at the bottom

JanSharp committed:
Make spell checker happy

JanSharp committed:
Reduce total amount tracked state

The less state that gets carried between function calls the more brain
  capacity is available to think about what the code is doing

JanSharp committed:
Further reduce the amount of shared state

Having shorter and isolated functions improves readability and with it
  maintainability. The removal of shared state also means there are
  fewer mistakes the programmer can make which would result in errors
  such as an invalid lexer `state` string, or not setting the delimiter
  variable when calling a function that expects it to be set
Remove first return value of parse_long_bracket_open. While it provided
  more descriptive code, it meant that the following code had to check
  if both variables were non nil, or work off of the assumption that the
  second return value is non nil when the first one is non nil. The
  latter is the more sane approach, however it still requires knowledge
  about when and how the function returns values. By removing the
  boolean and using the presence of the level as the boolean this
  duplication has been removed

JanSharp committed:
Move `---@plugin` parsing into the comment parser

JanSharp commented:
Progress update, with the same test file as before:
~27.8ms on 35db845 (upstream/next, next) Merge remote-tracking branch 'origin/pr/110' into next
~35.4ms on f5f1aab Merge branch 'next' into global-is-not-a-string-or-comment
~30.2ms on 7f3f6fa Change code ranges to use binary search
~29.2ms on ba123ae (global-is-not-a-string-or-comment) Move `---@plugin` parsing into the comment parser
So I can tell you already that it's going to be kept. I even had the idea to move some of the modules into the parser, because when profiling I've noticed that `string.match`/`string.gmatch` is the most expensive part in pretty much all the modules. But that'll be a separate PR most likely.

JanSharp committed:
Move module disable flags and related code down

RIP git history

JanSharp committed:
Move diff_finish_pos_to_diff_map clean up back up

Should have been done in 02fc3e6f3eff1a92f126cb9e095ed87e7aeca1f7

JanSharp committed:
Fix `=` for block strings/comments always being 0

This got broken in 9732bee754adc9a3871d7d4bf6d8c6ee81bac294

JanSharp committed:
Combine string/comment ranges into disabled flags

Move disabled check for the require module to the `require` word instead
  of inside of the string, because everything is disabled within strings
  One could argue that everything but the require module could be
  disabled within strings, however that's an obscure exception to make
  when simply moving the position to check is a very valid option
Normalize remote.call to also use the position of `remote` for disabled
  checks, same as remote.add_interface

JanSharp committed:
Stop `nil`ing `ranges` and `ranges_flags`

See added comment for details

JanSharp committed:
Remove comment checks in individual modules

Remove preceding_text capture where possible
  This is actually a 13% speedup in my test file

JanSharp committed:
Properly handle newlines for `disable-(next)-line`

Fix/handle adding zero width ranges, they do nothing

JanSharp committed:
Fix anchor being off by one

JanSharp committed:
Simplify binary search

Remove "try close to lower bound first" logic, because there is pattern
  across all kinds of source files which would guarantee that multiple
  matches for some module would be close together or far apart. Nor is
  there any guarantee for how many strings or comments would be in a
  file. All in all the logic does not add anything meaningful other than
  complexity. With its removal the binary search function was able to be
  cleaned up significantly

JanSharp committed:
Improve `parse_short_string` ever so slightly

JanSharp committed:
Remove `peek` and remove `string.sub` from `take`

There does not appear to be a measurable performance difference, however
  the fewer string creations the better. Plus `take` is smaller this way

JanSharp commented:
Alright, it's done. This is actually 10% +-5% faster than `next` branch, thanks to the removal of preceding_text captures from 2 modules (7ecc551).
Aside from that this is a big change and unfortunately due to potentially overlapping ranges the creation of the `ranges` and `ranges_flags` arrays has grown in complexity significantly. This is caused by `disable-next-line` and `disable-line`, since the ranges they affect could contain other ranges such as comments or strings. It's written with generated source files which have really long lines in mind, which is why all searches through the ranges array are done through binary searches, even when it's just searching through the last/current line.
The biggest issue right now is that even though I've tested it quite a bit, there's no guarantee that there are no bugs. It's too complex for that. Automated testing is on my list of things to add, but damn that'll be a lot of work :sweat_smile:
For the record I did check if anything else is inserting `--` which would combine with a `-` and cause syntax errors, but nope `global` was the only one.

Anyway, @penguinencounter and @justarandomgeek if you'd both like to git checkout the PR and use this version of the plugin in your mods folder (or similar) just to see if something obvious is broken that'd be awesome. With that I'm calling it good.

JanSharp commented:
I just noticed that the `package-lock.json` file got modified in 6fe5df0. I don't know enough to say if that's good/bad or doesn't matter.

JanSharp committed:
Fix cursor being off by 1 after `take`

Got broken in aa21e0a3723caae64f1b724f92989a6414f724cb

justarandomgeek commented:
> Anyway, @penguinencounter and @justarandomgeek if you'd both like to git checkout the PR and use this version of the plugin in your mods folder (or similar) just to see if something obvious is broken that'd be awesome. With that I'm calling it good.

Ah, excellent, i had been watching the progress initially but got distracted the last couple days, i'll check this out soon!

> I just noticed that the `package-lock.json` file got modified in [6fe5df0](https://github.com/justarandomgeek/vscode-factoriomod-debug/commit/6fe5df0d9dada39f8d6388c2c098bd20e756ae60). I don't know enough to say if that's good/bad or doesn't matter.

Mostly harmless. It gets touched any time any of the deps change and get pulled in. I'll almost certainly smash it again anyway.
